File: backend/metascraper/service.py
"""MetaScraper service layer — orchestrates store + diff + sheet sync, and
builds the read-model the dashboard consumes."""

from __future__ import annotations

from datetime import date, datetime, timezone
from typing import Any

from backend.metascraper import sheets, store
from backend.metascraper.diff import is_proven, longevity_ranks, merge_capture
from backend.metascraper.models import GROUP_LABELS, AdRecord, AppConfig, CaptureFile


def _today_iso() -> str:
    return datetime.now(timezone.utc).date().isoformat()


# ─── Config ───────────────────────────────────────────────────────────────────
def get_config() -> AppConfig:
    return store.load_config()


def save_config(data: dict[str, Any]) -> AppConfig:
    return store.save_config(AppConfig.from_dict(data))


def reset_config() -> AppConfig:
    return store.reset_config()


# ─── Ingest (the auto-fed endpoint) ───────────────────────────────────────────
def ingest_capture(capture: CaptureFile) -> dict[str, Any]:
    """Merge a capture into the store, persist, record the run, mirror to the
    sheet. Returns a summary the caller (and Claude-in-Chrome) can log."""
    today = _today_iso()
    before = store.load_ads()
    before_by_id = {a.library_id: a for a in before}

    merged = merge_capture(before, capture, today=today)
    store.save_ads(merged)

    # Tally what changed this run.
    merged_by_id = {a.library_id: a for a in merged}
    new_ids = [a.library_id for a in merged if a.status == "new" and a.library_id not in before_by_id]
    killed_ids = [
        a.library_id for a in merged
        if a.status == "killed" and before_by_id.get(a.library_id) is not None
        and before_by_id[a.library_id].status != "killed"
    ]
    summary = {
        "ingested": len(capture.ads),
        "store_size": len(merged),
        "new": len(new_ids),
        "killed_this_run": len(killed_ids),
        "running": sum(1 for a in merged if a.status == "running"),
    }
    store.record_capture(capture, summary)

    sheet_result = sheets.sync(merged)
    summary["sheet"] = sheet_result
    return summary


# ─── Read model for the dashboard ─────────────────────────────────────────────
def _ad_view(ad: AdRecord, rank: float | None) -> dict[str, Any]:
    data = ad.as_dict()
    data["proven"] = is_proven(ad.days_active)
    data["longevity_rank"] = rank
    return data


def get_ads_view() -> list[dict[str, Any]]:
    ads = store.load_ads()
    ranks = longevity_ranks(ads)
    view = [_ad_view(a, ranks.get(a.library_id)) for a in ads]
    # Longevity is the signal — default sort by days_active desc (nulls last).
    view.sort(key=lambda a: (a["days_active"] is not None, a["days_active"] or 0), reverse=True)
    return view


def get_summary() -> dict[str, Any]:
    """Per-niche summary cards + overall counts."""
    ads = store.load_ads()
    config = store.load_config()
    label = {n.id: n.label_jp for n in config.niches}
    group_of = {n.id: n.group for n in config.niches}

    per_niche: dict[str, dict[str, Any]] = {}
    for ad in ads:
        bucket = per_niche.setdefault(
            ad.niche_id,
            {
                "niche_id": ad.niche_id,
                "label_jp": label.get(ad.niche_id, ad.niche_id),
                "group": group_of.get(ad.niche_id, "other"),
                "group_label": GROUP_LABELS.get(group_of.get(ad.niche_id, "other"), "その他"),
                "active": 0,
                "killed": 0,
                "new": 0,
                "longest_days": None,
                "longest_ad": None,
            },
        )
        if ad.status == "killed":
            bucket["killed"] += 1
        else:
            bucket["active"] += 1
        if ad.status == "new":
            bucket["new"] += 1
        if ad.days_active is not None and (bucket["longest_days"] is None or ad.days_active > bucket["longest_days"]):
            bucket["longest_days"] = ad.days_active
            bucket["longest_ad"] = {
                "library_id": ad.library_id,
                "page_name": ad.page_name,
                "ad_library_url": ad.ad_library_url,
            }

    return {
        "totals": {
            "tracked": len(ads),
            "active": sum(1 for a in ads if a.status != "killed"),
            "killed": sum(1 for a in ads if a.status == "killed"),
            "proven": sum(1 for a in ads if is_proven(a.days_active)),
        },
        "niches": sorted(per_niche.values(), key=lambda n: n["active"], reverse=True),
        "last_capture": (store.list_captures() or [{}])[0].get("captured_date"),
    }


def get_diff_since_last() -> dict[str, Any]:
    """New + killed since the previous capture date — the weekly delta view."""
    captures = store.list_captures()
    if not captures:
        return {"since": None, "current": None, "new": [], "killed": []}

    current_date = captures[0]["captured_date"]
    prev_date = store.previous_capture_date(current_date)
    ads = store.load_ads()
    ranks = longevity_ranks(ads)

    new = [_ad_view(a, ranks.get(a.library_id)) for a in ads if a.first_seen == current_date]
    killed = [
        _ad_view(a, ranks.get(a.library_id))
        for a in ads
        if a.status == "killed" and prev_date is not None and a.last_seen == prev_date
    ]
    return {
        "since": prev_date,
        "current": current_date,
        "new": new,
        "killed": killed,
    }


def patch_ad(library_id: str, hook_category: str | None, notes: str | None) -> dict[str, Any] | None:
    ads = store.load_ads()
    target = next((a for a in ads if a.library_id == library_id), None)
    if target is None:
        return None
    if hook_category is not None:
        target.hook_category = hook_category or None
    if notes is not None:
        target.notes = notes or None
    store.update_ad(target)
    ranks = longevity_ranks(ads)
    return _ad_view(target, ranks.get(library_id))


def health() -> dict[str, Any]:
    from backend.config import settings

    return {
        "ingest_token_set": bool(settings.metascraper_ingest_token),
        "sheet_configured": sheets.is_configured(),
        "store_size": len(store.load_ads()),
        "last_capture": (store.list_captures() or [{}])[0].get("captured_date"),
    }

File: frontend/lib/metascraper/server/service.ts
import type { AppConfig } from "../types";
import type { CapturePayload, StoredAd } from "./diff";
import { isProven, longevityRanks, mergeCapture } from "./diff";
import { isConfigured, syncToSheet } from "./sheets";
import * as store from "./store";

const GROUP_LABELS: Record<string, string> = {
  surgical: "外科・整形系",
  injectable: "注入・注射系",
  skin: "美容皮膚科・肌治療",
  hair_removal: "脱毛",
  body: "痩身・ダイエット",
  hair_loss: "毛髪治療",
  mens: "メンズ",
  other: "その他・トレンド",
};

const HOOK_CATEGORIES = [
  "before_after", "price", "testimonial", "doctor_trust", "campaign", "other",
];

function todayIso(): string {
  return new Date().toISOString().slice(0, 10);
}

function adView(ad: StoredAd, rank: number | undefined) {
  return { ...ad, proven: isProven(ad.days_active), longevity_rank: rank ?? null };
}

// ─── Config ──────────────────────────────────────────────────────────────────

export async function getConfig(userId: string) {
  const config = await store.loadConfig(userId);
  return { config, groupLabels: GROUP_LABELS, hookCategories: HOOK_CATEGORIES };
}

export async function putConfig(userId: string, data: AppConfig) {
  const config = await store.saveConfig(userId, data);
  return { config };
}

export async function resetConfig(userId: string) {
  const config = await store.resetConfig(userId);
  return { config };
}

// ─── Ingest ──────────────────────────────────────────────────────────────────

export async function ingestCapture(userId: string, capture: CapturePayload) {
  const today = todayIso();
  const before = await store.loadAds(userId);
  const beforeById = new Map(before.map((a) => [a.library_id, a]));

  const merged = mergeCapture(before, capture, today);
  await store.saveAds(userId, merged);

  const newIds = merged
    .filter((a) => a.status === "new" && !beforeById.has(a.library_id))
    .map((a) => a.library_id);
  const killedIds = merged
    .filter(
      (a) =>
        a.status === "killed" &&
        beforeById.has(a.library_id) &&
        beforeById.get(a.library_id)!.status !== "killed",
    )
    .map((a) => a.library_id);

  const summary = {
    ingested: capture.ads.length,
    store_size: merged.length,
    new: newIds.length,
    killed_this_run: killedIds.length,
    running: merged.filter((a) => a.status === "running").length,
  };

  await store.recordCapture(userId, capture, summary);
  const sheet = await syncToSheet(merged);
  return { ...summary, sheet };
}

// ─── Dashboard reads ─────────────────────────────────────────────────────────

export async function getAdsView(userId: string) {
  const ads = await store.loadAds(userId);
  const ranks = longevityRanks(ads);
  const view = ads.map((a) => adView(a, ranks.get(a.library_id)));
  view.sort(
    (a, b) =>
      (b.days_active != null ? 1 : 0) - (a.days_active != null ? 1 : 0) ||
      (b.days_active ?? 0) - (a.days_active ?? 0),
  );
  return view;
}

export async function getSummary(userId: string) {
  const [ads, config] = await Promise.all([store.loadAds(userId), store.loadConfig(userId)]);
  const labelOf = Object.fromEntries(config.niches.map((n) => [n.id, n.label_jp]));
  const groupOf = Object.fromEntries(config.niches.map((n) => [n.id, n.group]));

  const perNiche: Record<string, {
    niche_id: string; label_jp: string; group: string; group_label: string;
    active: number; killed: number; new: number;
    longest_days: number | null; longest_ad: unknown | null;
  }> = {};

  for (const ad of ads) {
    const bucket = (perNiche[ad.niche_id] ??= {
      niche_id: ad.niche_id,
      label_jp: labelOf[ad.niche_id] ?? ad.niche_id,
      group: groupOf[ad.niche_id] ?? "other",
      group_label: GROUP_LABELS[groupOf[ad.niche_id] ?? "other"] ?? "その他",
      active: 0, killed: 0, new: 0,
      longest_days: null, longest_ad: null,
    });
    if (ad.status === "killed") bucket.killed++;
    else bucket.active++;
    if (ad.status === "new") bucket.new++;
    if (ad.days_active != null && (bucket.longest_days == null || ad.days_active > bucket.longest_days)) {
      bucket.longest_days = ad.days_active;
      bucket.longest_ad = { library_id: ad.library_id, page_name: ad.page_name, ad_library_url: ad.ad_library_url };
    }
  }

  const captures = await store.listCaptures(userId);
  return {
    totals: {
      tracked: ads.length,
      active: ads.filter((a) => a.status !== "killed").length,
      killed: ads.filter((a) => a.status === "killed").length,
      proven: ads.filter((a) => isProven(a.days_active)).length,
    },
    niches: Object.values(perNiche).sort((a, b) => b.active - a.active),
    last_capture: (captures[0]?.captured_date as string) ?? null,
  };
}

export async function getDiffSinceLast(userId: string) {
  const captures = await store.listCaptures(userId);
  if (!captures.length) return { since: null, current: null, new: [], killed: [] };

  const currentDate = captures[0].captured_date as string;
  const prevDate = await store.previousCaptureDate(userId, currentDate);
  const ads = await store.loadAds(userId);
  const ranks = longevityRanks(ads);

  return {
    since: prevDate,
    current: currentDate,
    new: ads.filter((a) => a.first_seen === currentDate).map((a) => adView(a, ranks.get(a.library_id))),
    killed: ads
      .filter((a) => a.status === "killed" && prevDate != null && a.last_seen === prevDate)
      .map((a) => adView(a, ranks.get(a.library_id))),
  };
}

export async function patchAd(
  userId: string,
  libraryId: string,
  patch: { hook_category?: string | null; notes?: string | null },
) {
  const ads = await store.loadAds(userId);
  const target = ads.find((a) => a.library_id === libraryId);
  if (!target) return null;
  if ("hook_category" in patch) target.hook_category = patch.hook_category ?? null;
  if ("notes" in patch) target.notes = patch.notes ?? null;
  await store.updateAd(userId, target);
  const ranks = longevityRanks(ads);
  return adView(target, ranks.get(libraryId));
}

export async function getCaptures(userId: string) {
  return store.listCaptures(userId);
}

export async function getHealth(userId: string) {
  const [ads, captures] = await Promise.all([store.loadAds(userId), store.listCaptures(userId)]);
  return {
    ingest_token_set: Boolean(process.env.METASCRAPER_INGEST_TOKEN),
    sheet_configured: isConfigured(),
    store_size: ads.length,
    last_capture: (captures[0]?.captured_date as string) ?? null,
  };
}
